### PR TITLE
#163499467 Save article draft 

### PR DESCRIPTION
**Pull request by MaggieChege**
**What does this PR do?**
- A user can create an article draft.
- A user can publish the article later.
- A user can view all the drafts he/she has.

**Description of Task to be completed?**
- A user can create an article draft and publish it later.

How should this be manually tested?
Clone the repo locally

cd ah-django-unchained

git checkout ft-save-article-draft-163499467

get started with the app as from the README section. Follow the sample .env file supplied to obtain the right variables

**SCREENSHOTS**

**Create an article draft**

image

**Get all drafts**

image

**Publish Article**

image

### CODE REVIEW

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -65,6 +66,22 @@ def get(self, request, slug, *args, **kwargs):
...
+
+class PublishArticle(UpdateAPIView):
+    """This class handles PUT requests."""
+    permission_classes = (IsAuthorOrReadOnly,)
+    queryset = Article.objects.filter(is_deleted=False, is_published=False)
+    serializer_class = PublishArticleSerializer
+    lookup_field = 'slug'
+
+
 class DeleteArticle(UpdateAPIView):
```
Comment (kwameasiago) [inline authors/apps/articles/views.py:96]: Add a class comment

**Review by d-kahara — APPROVED** [review]
Looks good to me

**Review by kwameasiago — APPROVED** [review]

File: authors/apps/articles/models.py [hunk 2]
```diff
@@ -19,6 +19,10 @@ class Article(models.Model):
     is_published = models.BooleanField(default=False)
     favorite = models.ManyToManyField(
         User, related_name='favorite', default=False)
+
+    favorite = models.ManyToManyField(User, related_name='favorite',
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/models.py:21]: favorites is duplicated

File: authors/apps/articles/serializers.py [hunk 3]
```diff
@@ -132,10 +137,15 @@ def to_representation(self, instance):
...
                 'createdAt': self.format_date(thread.createdAt),
                 'updatedAt': self.format_date(thread.updatedAt)
-            }for thread in instance.threads.all()
+            } for thread in instance.threads.all()
         ]
+
         thread_comment = super(
             CommentSerializer, self).to_representation(instance)
+
+        thread_comment = super(CommentSerializer, self).to_representation(
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/serializers.py:145]: duplicate

File: authors/apps/articles/urls.py [hunk 4]
```diff
@@ -6,9 +6,11 @@
                     FavoriteArticle, AverageRatingsAPIView,
                     CreateComment, CommentDetailsView,
                     CommentDelete, CommentsRetrieveUpdateDestroy,
-                    ReadTime, ShareArticleViaEmailApiView,
+                    ShareArticleViaEmailApiView,
                     ShareViaFacebook, ShareViaTwitter,
-                    LikeCommentApiView, DislikeCommentApiView
+                    LikeCommentApiView, DislikeCommentApiView,
+                    ReadTime, PublishArticle, GetDraft,
+                     PublishArticle, GetDraft
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/urls.py]: Duplicate imports

File: authors/apps/articles/urls.py [hunk 5]
```diff
@@ -58,4 +60,10 @@
          LikeCommentApiView.as_view(), name='comment_like'),
     path('articles/<slug>/comments/<int:id>/dislike',
          DislikeCommentApiView.as_view(), name='comment_dislike'), 
+
+    path('article/<slug>/publish/',
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/urls.py]: `articles/`

File: authors/apps/articles/views.py [hunk 6]
```diff
@@ -72,7 +72,24 @@ def get(self, request, slug, *args, **kwargs):
         return set_favorite_status(serializer, self.request.user.id)
 
 
+class GetDraft(ListAPIView):
+    """Get all Drafts"""
+    pagination_class = CustomPagination
+    permission_classes = (IsAuthorOrReadOnly,)
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/views.py]: I'll still be able to see drafts of other authors

File: authors/apps/articles/views.py [hunk 7]
```diff
@@ -72,7 +72,24 @@ def get(self, request, slug, *args, **kwargs):
...
+    queryset = Article.objects.filter(is_published=False, is_deleted=False)
+    serializer_class = GetArticleSerializer
+
+
+class PublishArticle(UpdateAPIView):
+    """This class handles PUT requests."""
+    permission_classes = (IsAuthorOrReadOnly,)
+    queryset = Article.objects.filter(is_deleted=False, is_published=False)
+    serializer_class = PublishArticleSerializer
+    lookup_field = 'slug'
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/views.py:93]: Let's return the published article

**Review by ignatiusukwuoma — COMMENTED** [review]
See these

File: authors/apps/articles/tests/test_create_article.py [hunk 8]
```diff
@@ -102,17 +105,31 @@ def post_article_articles(self, data):
...
 
-    def test_fetch_all_articles(self):
+    def test_fetch_all_published_articles(self):
         """Tests to fetch all articles"""
         token = self.signup_user_one()
+        # create article
         self.create_article(self.create_article_data, token)
         self.create_article(self.create_article_data2, token)
+        # publish articles
+        self.client.put(reverse('articles:publish_article',
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_create_article.py:118]: Assert the value of `is_published` is as expected before and after publishing

File: authors/apps/articles/tests/test_create_article.py [hunk 9]
```diff
@@ -124,14 +141,15 @@ def test_fetch_all_articles(self):
...
+    def test_create_article_draft(self):
+        """Test to create article draft"""
         token = self.signup_user_one()
         del self.create_article_data['images']
         response = self.client.post(self.article_listcreate,
                                     self.create_article_data,
                                     format='json',
-                                    HTTP_AUTHORIZATION='token {}'.format(token))
+                                    HTTP_AUTHORIZATION='token {}'.format(
+                                        token))
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_create_article.py:156]: Also assert the value of is_published, that's what makes it a draft

File: authors/apps/articles/urls.py [hunk 10]
```diff
@@ -58,4 +60,10 @@
...
+         PublishArticle.as_view(),
+         name="publish_article"),
+    path('article/draft/', GetDraft.as_view(),
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/urls.py]: `articles/drafts/`

File: authors/apps/articles/tests/test_create_article.py [hunk 11]
```diff
@@ -102,17 +107,33 @@ def post_article_articles(self, data):
...
+                        format='json',
+                        HTTP_AUTHORIZATION='token {}'.format(token))
+        self.client.put(reverse('articles:publish_article',
+                                kwargs={'slug': 'the-beginning'}),
+                        self.publish_data,
+                        format='json',
+                        HTTP_AUTHORIZATION='token {}'.format(token))
         response = self.client.get(self.article_listcreate, format='json')
         self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data['results'][0]['is_published'], True)
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_create_article.py]: assertTrue

File: authors/apps/articles/tests/test_create_article.py [hunk 12]
```diff
@@ -102,17 +107,33 @@ def post_article_articles(self, data):
...
 
-    def test_fetch_all_articles(self):
+    def test_fetch_all_published_articles(self):
         """Tests to fetch all articles"""
         token = self.signup_user_one()
+        # create article
         self.create_article(self.create_article_data, token)
         self.create_article(self.create_article_data2, token)
+        # publish articles
+        self.assertNotEqual(self.create_article_data['is_published'], True)
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/tests/test_create_article.py:117]: assertFalse

File: authors/apps/articles/views.py [hunk 13]
```diff
@@ -74,7 +74,24 @@ def get(self, request, slug, *args, **kwargs):
         return set_favorite_status(serializer, self.request.user.id)
 
 
+class GetDraft(ListAPIView):
+    """Get all Drafts"""
+    pagination_class = CustomPagination
+    permission_classes = (IsAuthenticated,)
```
Comment (ignatiusukwuoma) [inline authors/apps/articles/views.py]: You can create an `isAuthor` class for this... so only an author can see his drafts

**Review by ignatiusukwuoma — APPROVED** [review]

**Pull request merged**